### ISSUE TITLE
Lora

### ISSUE THREAD

**Issue by 2kxx** [zhiyuanyou/DeQA-Score#6]
有用Lora在KonIQ单独训练时的sh脚本吗？Lora训练时未设置eval集，是直接默认取最后训练结果作为最优来测试？

**Comment by zhiyuanyou** [comment]
在训练脚本 https://github.com/zhiyuanyou/DeQA-Score/blob/main/scripts/train_lora.sh 中，修改 --data_paths 和 --data_weights 即可实现 KonIQ 单独训练。

**Comment by zhiyuanyou** [comment]
对的，直接取最后的训练结果来测试。

**Comment by 2kxx** [comment]
嗯嗯，这个我知道的，我的意思是超参数的设置呢，比如学习率，最后测试的checkpoint这些有没有什么变化？

**Comment by zhiyuanyou** [comment]
按理说最后一个checkpoint不是最优的，但我们测试下来效果已经很好了，所以没有通过验证集选择checkpoint。

**Comment by 2kxx** [comment]
> 按理说最后一个checkpoint不是最优的，但我们测试下来效果已经很好了，所以没有通过验证集选择checkpoint。

okk感谢，我是发现Q-Align在KonIQ单独训练时的学习率比你们的小了10倍，然后他们并没有给出Lora的性能差异，所以想问下你们的设计

**Comment by zhiyuanyou** [comment]
我们release的模型是在3个数据集混合训练的。如果是KonIQ单独训练，可以自己再调一下参数。一次训练时间很短，应该1-2小时就训好了。

**Comment by 2kxx** [comment]
还想问下代码中的num_vision_tokens是这么来的，为什么在描述task的情况下就是维度相等、一一对齐，可以直接计算txt loss呢：
`num_vision_tokens = logits.shape[1] - labels.shape[1]
        idx_level_logit = idx_level_label + num_vision_tokens - 1
        logits_level_ids = logits[
            torch.arange(batch_size), idx_level_logit
        ].contiguous()  # [B, V]`

**Comment by zhiyuanyou** [comment]
在两个函数中出现了num_vision_tokens。

第一个是get_score():
num_vision_tokens = outputs.logits.shape[1] - item["labels"].shape[1]
在这里，logits是包括了vision + text的全部输出，labels是人构建的labels，只包括了text，所以减去之后，num_vision_tokens是vision token的数量。

第二个是softkl_loss():
num_vision_tokens = logits.shape[1] - labels.shape[1]
在这里，logits是包括了vision + text的全部输出，labels是经过了prepare_inputs_for_generation()的labels，已经pad了IGNORE_INDEX，因此长度是一致的，所以减去之后，num_vision_tokens是0。


你可以验证一下，第一个不是0，第二个是0。如果不是的话，可以反馈给我。

**Comment by 2kxx** [comment]
> 在两个函数中出现了num_vision_tokens。
> 
> 第一个是get_score(): num_vision_tokens = outputs.logits.shape[1] - item["labels"].shape[1] 在这里，logits是包括了vision + text的全部输出，labels是人构建的labels，只包括了text，所以减去之后，num_vision_tokens是vision token的数量。
> 
> 第二个是softkl_loss(): num_vision_tokens = logits.shape[1] - labels.shape[1] 在这里，logits是包括了vision + text的全部输出，labels是经过了prepare_inputs_for_generation()的labels，已经pad了IGNORE_INDEX，因此长度是一致的，所以减去之后，num_vision_tokens是0。
> 
> 你可以验证一下，第一个不是0，第二个是0。如果不是的话，可以反馈给我。

嗯嗯，确实是这样的，十分感谢你的耐心回答😀

**Comment by 2kxx** [comment]
还想请问下你们注释中的lm_head：
    for n, p in model.named_parameters():
        if training_args.lora_enable:
            p.requires_grad = True if "lora_" in n else False
        else:
            p.requires_grad = True
        # if "lm_head" in n:
        #     print(n)
        #     p.requires_grad = True
这里的lm_head是在get_peft_model之后定义的，所以如果我去除，是开启的lm_head的lora形式吗？
我如果要开启原本的，可以通过以下代码吗？
model.lm_head.requires_grad_(True)
    for n, p in model.lm_head.named_parameters():
        p.requires_grad = True
